### PR TITLE
Workaround to handle DCM (Dual-Chip Module) package for Redfish

### PR DESCRIPTION
**Pull request by RameshIyyar**
Created this PR to address the below comments.

- Enabled the Hardware Isolation (aka Guard) feature for Core and Memory #77 (comment)

**Note:** 
- This workaround (Refer commit message for more details) patch is required to enable the manual guard feature. 
- This PR is created on top of https://github.com/ibm-openbmc/bmcweb/pull/95 so, please ignore the first commit in this PR.

### CODE REVIEW

**Comment by RameshIyyar** [comment]
> * This PR is created on top of [Implement SubProcessors for processor core #95](https://github.com/ibm-openbmc/bmcweb/pull/95) so, please ignore the first commit in this PR.

Rebased this PR since #95 was merged

File: redfish-core/lib/processor.hpp [hunk 1]
```diff
@@ -47,6 +47,51 @@ constexpr std::array<const char*, 2> processorInterfaces = {
...
+ */
+inline bool
+    isProcObjectMatched(const std::string& processorId,
+                        const sdbusplus::message::object_path& objectPath)
+{
+    bool isMatched = false;
+    if (processorId.find("dcm") != std::string::npos)
+    {
+        std::string procId{processorId};
+        if (auto pos = procId.find('-'); pos != std::string::npos)
```
Comment (gtmills) [inline redfish-core/lib/processor.hpp:76]: Would this be separated on the line above to match bmcweb style Also to use std::size_t

like

https://github.com/openbmc/bmcweb/blob/97a056ac02e0f246a24d1a4353a3d24b109ccf2e/include/authorization.hpp#L57

File: redfish-core/lib/processor.hpp [hunk 2]
```diff
@@ -47,6 +47,51 @@ constexpr std::array<const char*, 2> processorInterfaces = {
...
+        {
+            procId.replace(pos, 1, "/");
+            if (objectPath.str.find(procId) != std::string::npos)
```
Comment (gtmills) [inline redfish-core/lib/processor.hpp:79]: Imprecise matching 
Common error 12. Can you use precise matching?

**Comment by gtmills** [comment]
Looks okay. Couple minor things.

**Comment by RameshIyyar** [comment]
> Looks okay. Couple minor things.

Addressed those comments, please take a look.

File: redfish-core/lib/processor.hpp [hunk 3]
```diff
@@ -47,6 +47,55 @@ constexpr std::array<const char*, 2> processorInterfaces = {
...
+{
+    bool isMatched = false;
+    if (processorId.find("dcm") != std::string::npos)
+    {
+        std::size_t delimiterPos = processorId.find('-');
+        if (delimiterPos != std::string::npos)
+        {
+            std::string procParent{processorId.substr(0, delimiterPos)};
+            std::string procId{
+                processorId.substr(delimiterPos + 1, processorId.length())};
```
Comment (gtmills) [inline redfish-core/lib/processor.hpp:80]: Don't we need to check the size of processorId ?

https://github.com/openbmc/bmcweb/blob/master/COMMON_ERRORS.md#4-walking-off-the-end-of-a-string

Also, I know this is the same but I don't see "std::string procId{" anywhere else in bmcweb..

https://github.com/openbmc/bmcweb/search?q=substr

    std::string procId = processorId.substr(delimiterPos + 1, processorId.length())

**Comment by RameshIyyar** [comment]
@gtmills please review

File: redfish-core/lib/processor.hpp (hunk 3 above)
Comment (gtmills) [reply redfish-core/lib/processor.hpp]: > Not required, are you seeing any issue with that?

What if your processorId looks like "dcmN-" ? 
We have these checks not for the good case but for the bad case so bmcweb doesn't throw an exception 
Still think we should add a check 

ttps://github.com/openbmc/bmcweb/blob/master/COMMON_ERRORS.md#4-walking-off-the-end-of-a-string is this exact case

Comment (RameshIyyar) [reply redfish-core/lib/processor.hpp]: > What if your processorId looks like "dcmN-" ?

**procId** will be empty in that case and it won't match in below `if` and bmcweb will return below errors:
```
if ((objectPath.parent_path().filename() == procParent) &&
    (objectPath.filename() == procId))
{   
    isMatched = true;
}
```

Error:
```
curl -k -H "X-Auth-Token: $bmc_token" -X GET https://${bmc}/redfish/v1/Systems/system/Processors/dcm0-
{
  "error": {
    "@Message.ExtendedInfo": [
      {
        "@odata.type": "#Message.v1_1_1.Message",
        "Message": "The requested resource of type Processor named dcm0- was not found.",
        "MessageArgs": [
          "Processor",
          "dcm0-"
        ],
        "MessageId": "Base.1.8.1.ResourceNotFound",
        "MessageSeverity": "Critical",
        "Resolution": "Provide a valid resource identifier and resubmit the request."
      }
    ],
    "code": "Base.1.8.1.ResourceNotFound",
    "message": "The requested resource of type Processor named dcm0- was not found."
  }
}
```

**Review by gtmills — APPROVED** [review]

File: redfish-core/lib/processor.hpp (hunk 3 above)
Comment (gtmills) [reply redfish-core/lib/processor.hpp]: Still think we need to check here, but not sure I have the energy to fight here

**Pull request merged**